File: parser.php
<?php

require_once __DIR__ . '/database.php';

function parseHtmlAndSave(string $contents): int {
    $dom = new DOMDocument();
    @$dom->loadHTML('<?xml encoding="UTF-8">' . $contents, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD);
    $count = 0;

    $substitutions = loadJson('substitutions');

    $tables = $dom->getElementsByTagName('table');

    foreach ($tables as $table) {
        $rows = $table->getElementsByTagName('tr');
        if ($rows->length < 3) continue;

        $dateHeader = trim($rows->item(0)->textContent);

        $dateStr = null;
        if (strpos($dateHeader, 'Dzień:') !== false) {
            $parts = explode('Dzień:', $dateHeader);
            $afterDay = trim($parts[1] ?? '');
            $beforeParen = explode('(', $afterDay);
            $dateStr = trim($beforeParen[0]);
        } elseif (strpos($dateHeader, 'Okres:') !== false) {
            continue;
        }

        if (!$dateStr) continue;

        $dateObj = DateTime::createFromFormat('d.m.Y', $dateStr);
        if (!$dateObj) continue;
        $dateFormatted = $dateObj->format('Y-m-d');

        for ($i = 2; $i < $rows->length; $i++) {
            $row = $rows->item($i);
            $cells = [];
            foreach ($row->childNodes as $child) {
                if ($child->nodeName === 'td' || $child->nodeName === 'th') {
                    $cells[] = trim($child->textContent);
                }
            }

            if (count($cells) !== 8) continue;

            try {
                $lessonText = $cells[0];
                if (!$lessonText || strpos($lessonText, ',') === false) continue;

                $lessonParts = explode(',', $lessonText);
                $lessonNum = (int)trim($lessonParts[0]);
                if ($lessonNum <= 0) continue;

                $lessonStartTime = null;
                $lessonEndTime = null;
                if (isset($lessonParts[1])) {
                    $timeStr = trim($lessonParts[1]);
                    $timeParts = explode('-', $timeStr);
                    if (count($timeParts) === 2) {
                        $lessonStartTime = trim($timeParts[0]);
                        $lessonEndTime = trim($timeParts[1]);
                    }
                }

                $originalTeacherName = trim($cells[1]);
                $classNameRaw = trim($cells[2]);
                $classParts = explode('|', $classNameRaw);
                $className = trim($classParts[0] ?? '');
                $classGroup = isset($classParts[1]) ? trim($classParts[1]) : null;

                $subject = trim($cells[3]);
                $classroomName = trim($cells[4]);
                $zastepcaRaw = trim($cells[5]);
                $cause = trim($cells[6]);
                $note = trim($cells[7]);

                if ($classGroup) {
                    if (preg_match('/\d+(?:\/\d+)?/', $classGroup, $m)) {
                        $groupValue = $m[0];
                        $groupText = "Grupa: {$groupValue}";
                        $note = $note ? "{$note}. {$groupText}" : $groupText;
                    }
                }

                if (!$className || !$subject) continue;

                $originalTeacherName = ($originalTeacherName && !in_array($originalTeacherName, ['-', 'wakat'])) ? $originalTeacherName : null;
                $classroomName = ($classroomName && $classroomName !== '-') ? $classroomName : null;
                $cause = ($cause && $cause !== '-') ? $cause : '';
                $note = ($note && $note !== '-') ? $note : '';

                $statusKeywords = ['uczniowie', 'okienko', 'zastępstwo', 'wakat'];
                $isStatusMessage = false;
                foreach ($statusKeywords as $kw) {
                    if (stripos($zastepcaRaw, $kw) !== false) {
                        $isStatusMessage = true;
                        break;
                    }
                }

                $substituteTeacherName = null;
                if ($zastepcaRaw) {
                    if (!$isStatusMessage) {
                        $substituteTeacherName = $zastepcaRaw;
                    } else {
                        if ($note) {
                            if (stripos($note, $zastepcaRaw) === false) {
                                $note = "{$zastepcaRaw}. {$note}";
                            }
                        } else {
                            $note = $zastepcaRaw;
                        }
                    }
                }

                [$classId] = getOrCreateClass($className);

                $classroomId = null;
                if ($classroomName) {
                    [$classroomId] = getOrCreateClassroom($classroomName);
                }

                $originalTeacherId = null;
                if ($originalTeacherName) {
                    [$originalTeacherId] = getOrCreateTeacher($originalTeacherName);
                }

                $substituteTeacherId = null;
                if ($substituteTeacherName) {
                    [$substituteTeacherId] = getOrCreateTeacher($substituteTeacherName);
                }

                $substitutions[] = [
                    'id' => nextId($substitutions),
                    'date' => $dateFormatted,
                    'lesson_number' => $lessonNum,
                    'lesson_start_time' => $lessonStartTime,
                    'lesson_end_time' => $lessonEndTime,
                    'original_teacher_id' => $originalTeacherId,
                    'substitute_teacher_id' => $substituteTeacherId,
                    'class_id' => $classId,
                    'classroom_id' => $classroomId,
                    'subject' => $subject,
                    'cause' => $cause,
                    'note' => $note,
                ];
                $count++;
            } catch (Exception $e) {
                continue;
            }
        }
    }

    saveJson('substitutions', $substitutions);
    return $count;
}

function parsePlanXmlAndSave(string $contents): int {
    $xml = @simplexml_load_string($contents);
    if (!$xml) {
        throw new Exception("Error parsing XML");
    }

    $count = 0;
    $lessons = loadJson('lessons');

    foreach ($xml->xpath('.//lesson') as $lessonElem) {
        try {
            $dayOfWeek = (int)($lessonElem['day'] ?? 1);
            $lessonNumber = (int)($lessonElem['hour'] ?? 1);
            $subject = (string)($lessonElem['subject'] ?? '');
            $teacherName = (string)($lessonElem['teacher'] ?? '');
            $className = (string)($lessonElem['class'] ?? '');
            $classroomName = (string)($lessonElem['classroom'] ?? '') ?: null;

            if (!$teacherName || !$className || !$subject) continue;

            [$teacherId] = getOrCreateTeacher($teacherName);
            [$classId] = getOrCreateClass($className);

            $classroomId = null;
            if ($classroomName) {
                [$classroomId] = getOrCreateClassroom($classroomName);
            }

            $lessons[] = [
                'id' => nextId($lessons),
                'teacher_id' => $teacherId,
                'class_id' => $classId,
                'classroom_id' => $classroomId,
                'subject' => $subject,
                'day_of_week' => $dayOfWeek,
                'lesson_number' => $lessonNumber,
            ];
            $count++;
        } catch (Exception $e) {
            continue;
        }
    }

    saveJson('lessons', $lessons);
    return $count;
}

function parsePlanHtmlFolderAndSave(array $htmlFiles): int {
    $count = 0;
    $lessons = loadJson('lessons');

    // Build name maps from lista.html
    $classNameMap = [];   // "o36.html" => "4TMe 4technik mechatronik"
    $teacherNameMap = []; // "n9.html"  => ["name" => "A.Skoczek (SK)", "short" => "SK"]

    foreach ($htmlFiles as $filename => $content) {
        if (basename($filename) === 'lista.html') {
            $dom = new DOMDocument();
            @$dom->loadHTML('<?xml encoding="UTF-8">' . $content, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD);
            foreach ($dom->getElementsByTagName('a') as $link) {
                $base = basename($link->getAttribute('href'));
                $text = trim($link->textContent);
                if (preg_match('/^o\d+\.html$/i', $base)) {
                    $classNameMap[$base] = $text;
                } elseif (preg_match('/^n\d+\.html$/i', $base)) {
                    $short = '';
                    if (preg_match('/\((\w+)\)\s*$/', $text, $m)) {
                        $short = $m[1];
                    }
                    $teacherNameMap[$base] = ['name' => $text, 'short' => $short];
                }
            }
            break;
        }
    }

    // Extract lesson hours from any class plan file and save to hours.json
    $hoursExtracted = false;
    foreach ($htmlFiles as $filename => $contents) {
        $base = basename($filename);
        if (!preg_match('/^o\d+\.html$/i', $base)) continue;

        $domH = new DOMDocument();
        @$domH->loadHTML('<?xml encoding="UTF-8">' . $contents, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD);
        $xpathH = new DOMXPath($domH);
        $tables = $xpathH->query('//table[@class="tabela"]');
        if ($tables->length === 0) continue;
        $planTable = $tables->item(0);
        $rows = $planTable->getElementsByTagName('tr');
        $hours = [];
        for ($ri = 1; $ri < $rows->length; $ri++) {
            $row = $rows->item($ri);
            $cells = [];
            foreach ($row->childNodes as $child) {
                if ($child->nodeName === 'td' || $child->nodeName === 'th') {
                    $cells[] = $child;
                }
            }
            if (count($cells) < 2) continue;
            $nrText = trim($cells[0]->textContent);
            preg_match('/\d+/', $nrText, $nm);
            if (!isset($nm[0])) continue;
            $nr = (int)$nm[0];
            $godzText = trim($cells[1]->textContent);
            // Format: " 7:15- 8:00"
            if (preg_match('/(\d+:\d+)\s*-\s*(\d+:\d+)/', $godzText, $tm)) {
                $hours[] = ['nr' => $nr, 'start' => $tm[1], 'end' => $tm[2]];
            }
        }
        if (!empty($hours)) {
            saveJson('hours', $hours);
            $hoursExtracted = true;
            break;
        }
    }

    // Process only class plan files (o*.html)
    foreach ($htmlFiles as $filename => $contents) {
        $base = basename($filename);
        if (!preg_match('/^o\d+\.html$/i', $base)) continue;

        try {
            // Resolve class name
            $fullClassName = $classNameMap[$base] ?? null;
            if (!$fullClassName) {
                $dom = new DOMDocument();
                @$dom->loadHTML('<?xml encoding="UTF-8">' . $contents, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD);
                $titles = $dom->getElementsByTagName('title');
                if ($titles->length > 0 && preg_match('/[-–]\s*(\S+)/', $titles->item(0)->textContent, $m)) {
                    $fullClassName = $m[1];
                }
            }
            if (!$fullClassName) continue;

            // Use short class name (first word) for consistency with substitutions
            $className = explode(' ', $fullClassName)[0];
            [$classId] = getOrCreateClass($className);

            $dom = new DOMDocument();
            @$dom->loadHTML('<?xml encoding="UTF-8">' . $contents, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD);
            $xpath = new DOMXPath($dom);

            // Find timetable (table.tabela)
            $tables = $xpath->query('//table[@class="tabela"]');
            if ($tables->length === 0) continue;
            $planTable = $tables->item(0);

            $rows = $planTable->getElementsByTagName('tr');

            for ($ri = 1; $ri < $rows->length; $ri++) {
                $row = $rows->item($ri);

                $cells = [];
                foreach ($row->childNodes as $child) {
                    if ($child->nodeName === 'td' || $child->nodeName === 'th') {
                        $cells[] = $child;
                    }
                }

                // Columns: 0=Nr, 1=Godz, 2=Mon, 3=Tue, 4=Wed, 5=Thu, 6=Fri
                if (count($cells) < 7) continue;

                $lessonNumText = trim($cells[0]->textContent);
                preg_match('/\d+/', $lessonNumText, $nm);
                if (!isset($nm[0])) continue;
                $lessonNum = (int)$nm[0];
                if ($lessonNum <= 0) continue;

                for ($di = 0; $di < 5; $di++) {
                    $dayOfWeek = $di + 1;
                    $cell = $cells[$di + 2];

                    // Each <span class="p"> is one lesson entry
                    foreach ($xpath->query('.//span[@class="p"]', $cell) as $subjectSpan) {
                        $subject = trim($subjectSpan->textContent);
                        if (!$subject) continue;

                        // Teacher and classroom are siblings within the same container
                        $container = $subjectSpan->parentNode;
                        $teacherNodes = $xpath->query('.//a[@class="n"]', $container);
                        $classroomNodes = $xpath->query('.//a[@class="s"]', $container);

                        if ($teacherNodes->length === 0) continue;

                        $teacherAnchor = $teacherNodes->item(0);
                        $teacherShort = trim($teacherAnchor->textContent);
                        $teacherFile = basename($teacherAnchor->getAttribute('href'));

                        $classroomName = $classroomNodes->length > 0
                            ? trim($classroomNodes->item(0)->textContent)
                            : '';

                        // Resolve full teacher name from lista map
                        $teacherName = $teacherShort;
                        $teacherShortName = $teacherShort;
                        if ($teacherFile && isset($teacherNameMap[$teacherFile])) {
                            $teacherName = $teacherNameMap[$teacherFile]['name'];
                            $teacherShortName = $teacherNameMap[$teacherFile]['short'] ?: $teacherShort;
                        }

                        [$teacherId] = getOrCreateTeacher($teacherName, $teacherShortName);

                        $classroomId = null;
                        if ($classroomName !== '' && $classroomName !== '-') {
                            [$classroomId] = getOrCreateClassroom($classroomName);
                        }

                        $lessons[] = [
                            'id' => nextId($lessons),
                            'teacher_id' => $teacherId,
                            'class_id' => $classId,
                            'classroom_id' => $classroomId,
                            'subject' => $subject,
                            'day_of_week' => $dayOfWeek,
                            'lesson_number' => $lessonNum,
                        ];
                        $count++;
                    }
                }
            }
        } catch (Exception $e) {
            continue;
        }
    }

    saveJson('lessons', $lessons);
    return $count;
}

File: parser.py
"""
Parser module for handling HTML and XML imports of substitutions and schedules.
"""

from bs4 import BeautifulSoup
from xml.etree import ElementTree as ET
from datetime import datetime
import re
import crud, models, schemas
from sqlalchemy.orm import Session


def parse_html_and_save(contents: bytes, db: Session) -> int:
    """
    Parse HTML file containing substitutions and save to database.
    Handles multi-table format with one table per day.
    
    Args:
        contents: Raw HTML file contents
        db: Database session
        
    Returns:
        Number of substitutions processed
    """
    try:
        soup = BeautifulSoup(contents, 'html.parser')
        count = 0
        
        # Find all tables
        tables = soup.find_all('table')
        
        for table in tables:
            rows = table.find_all('tr')
            if len(rows) < 3:
                continue
            
            # First row contains the date info (e.g., "Dzień: 02.03.2026 (pon.)")
            date_header = rows[0].get_text(strip=True)
            
            # Extract date from header like "Dzień: 02.03.2026 (pon.)"
            date_str = None
            if 'Dzień:' in date_header:
                # Format: "Dzień: 02.03.2026 (pon.)"
                parts = date_header.split('Dzień:')[1].strip().split('(')[0].strip()
                date_str = parts
            elif 'Okres:' in date_header:
                # Skip period headers, get date from next table
                continue
            
            if not date_str:
                continue
            
            # Parse date
            try:
                date_obj = datetime.strptime(date_str, "%d.%m.%Y").date()
            except ValueError:
                continue
            
            # Second row contains headers, skip it
            # Data rows start from row 2
            for row in rows[2:]:
                cells = row.find_all(['td', 'th'])
                if len(cells) != 8:
                    continue
                
                try:
                    # Column mapping (exact from HTML):
                    # 0: Lekcja (lesson number with time, e.g., "1, 07:15-08:00")
                    # 1: Nauczyciel/wakat (original teacher - who is absent)
                    # 2: Oddział (class)
                    # 3: Przedmiot (subject)
                    # 4: Sala (classroom)
                    # 5: Zastępca (substitute - can be teacher name or status message)
                    # 6: Powód (cause - reason for absence)
                    # 7: Uwagi (notes - additional info)
                    
                    lesson_text = cells[0].get_text(strip=True) if cells[0] else ""
                    if not lesson_text or ',' not in lesson_text:
                        continue
                    
                    # Extract lesson number from "1, 07:15-08:00"
                    try:
                        lesson_num = int(lesson_text.split(',')[0].strip())
                    except (ValueError, IndexError):
                        continue
                    
                    original_teacher_name = (cells[1].get_text(strip=True) if cells[1] else "").strip()
                    class_name_raw = (cells[2].get_text(strip=True) if cells[2] else "").strip()
                    # Extract main class name and group if exists
                    # E.g., "2bme|2/2" -> class_name="2bme", group="2/2"
                    class_parts = class_name_raw.split('|')
                    class_name = class_parts[0].strip() if class_parts else ""
                    class_group = class_parts[1].strip() if len(class_parts) > 1 else None
                    
                    subject = (cells[3].get_text(strip=True) if cells[3] else "").strip()
                    classroom_name = (cells[4].get_text(strip=True) if cells[4] else "").strip()
                    zastepca_raw = (cells[5].get_text(strip=True) if cells[5] else "").strip()
                    cause = (cells[6].get_text(strip=True) if cells[6] else "").strip()
                    note = (cells[7].get_text(strip=True) if cells[7] else "").strip()
                    
                    # Add class group to notes if it exists
                    if class_group:
                        # Extract only numeric part from group (e.g., "jn1" -> "1", "2/2" -> "2/2")
                        numeric_group = re.search(r'\d+(?:/\d+)?', class_group)
                        if numeric_group:
                            group_text = f"Grupa: {numeric_group.group()}"
                            if note:
                                note = f"{note}. {group_text}"
                            else:
                                note = group_text
                    
                    # Validate required fields
                    if not class_name or not subject:
                        continue
                    
                    # Normalize empty values to None where appropriate
                    original_teacher_name = original_teacher_name if original_teacher_name and original_teacher_name not in ['-', 'wakat'] else None
                    classroom_name = classroom_name if classroom_name and classroom_name not in ['-'] else None
                    cause = cause if cause and cause not in ['-'] else ""
                    note = note if note and note not in ['-'] else ""
                    
                    # Determine if zastępca column contains a real teacher name or a status message
                    status_keywords = ['uczniowie', 'okienko', 'zastępstwo', 'wakat']
                    is_status_message = any(keyword.lower() in zastepca_raw.lower() for keyword in status_keywords)
                    
                    # Extract substitute teacher
                    substitute_teacher_name = None
                    if zastepca_raw:
                        if not is_status_message:
                            # It's a real teacher name
                            substitute_teacher_name = zastepca_raw
                        else:
                            # It's a status message - add to notes
                            if note:
                                if zastepca_raw.lower() not in note.lower():
                                    note = f"{zastepca_raw}. {note}".strip()
                            else:
                                note = zastepca_raw
                    
                    # Get or create school class
                    school_class = crud.get_class_by_name(db, class_name)
                    if not school_class:
                        school_class = crud.create_class(db, schemas.SchoolClassCreate(name=class_name))
                    
                    # Get or create classroom if specified
                    classroom_id = None
                    if classroom_name:
                        classroom = crud.get_classroom_by_name(db, classroom_name)
                        if not classroom:
                            classroom = crud.create_classroom(db, schemas.ClassroomCreate(name=classroom_name))
                        classroom_id = classroom.id
                    
                    # Get or create original teacher (the absent one)
                    original_teacher_id = None
                    if original_teacher_name:
                        original_teacher = crud.get_teacher_by_name(db, original_teacher_name)
                        if not original_teacher:
                            short_name = original_teacher_name[:10]
                            original_teacher = crud.create_teacher(
                                db, 
                                schemas.TeacherCreate(name=original_teacher_name, short_name=short_name)
                            )
                        original_teacher_id = original_teacher.id
                    
                    # Get or create substitute teacher (the replacement)
                    substitute_teacher_id = None
                    if substitute_teacher_name:
                        substitute_teacher = crud.get_teacher_by_name(db, substitute_teacher_name)
                        if not substitute_teacher:
                            short_name = substitute_teacher_name[:10]
                            substitute_teacher = crud.create_teacher(
                                db,
                                schemas.TeacherCreate(name=substitute_teacher_name, short_name=short_name)
                            )
                        substitute_teacher_id = substitute_teacher.id
                    
                    # Create substitution record
                    substitution = schemas.SubstitutionCreate(
                        date=date_obj,
                        lesson_number=lesson_num,
                        original_teacher_id=original_teacher_id,
                        substitute_teacher_id=substitute_teacher_id,
                        class_id=school_class.id,
                        classroom_id=classroom_id,
                        subject=subject,
                        cause=cause,
                        note=note
                    )
                    
                    crud.create_substitution(db, substitution)
                    count += 1
                    
                except Exception as e:
                    continue
        
        return count
        
    except Exception as e:
        raise Exception(f"Error parsing HTML: {str(e)}")


def parse_plan_xml_and_save(contents: bytes, db: Session) -> int:
    """
    Parse XML file containing school schedule and save to database.
    
    Args:
        contents: Raw XML file contents
        db: Database session
        
    Returns:
        Number of lessons processed
    """
    try:
        root = ET.fromstring(contents)
        count = 0
        
        # Common XML structure for Optivum exported plans
        # Adjust XPath based on actual XML structure
        
        # Process lessons/classes
        for lesson_elem in root.findall('.//lesson'):
            try:
                # Extract attributes - adjust based on actual XML structure
                day_of_week = int(lesson_elem.get('day', 1))
                lesson_number = int(lesson_elem.get('hour', 1))
                subject = lesson_elem.get('subject', '')
                
                teacher_name = lesson_elem.get('teacher', '')
                class_name = lesson_elem.get('class', '')
                classroom_name = lesson_elem.get('classroom')
                
                if not (teacher_name and class_name and subject):
                    continue
                
                # Get or create teacher
                teacher = crud.get_teacher_by_name(db, teacher_name)
                if not teacher:
                    short_name = teacher_name[:10]
                    teacher = crud.create_teacher(
                        db,
                        schemas.TeacherCreate(name=teacher_name, short_name=short_name)
                    )
                
                # Get or create school class
                school_class = crud.get_class_by_name(db, class_name)
                if not school_class:
                    school_class = crud.create_class(db, schemas.SchoolClassCreate(name=class_name))
                
                # Get or create classroom
                classroom_id = None
                if classroom_name:
                    classroom = crud.get_classroom_by_name(db, classroom_name)
                    if not classroom:
                        classroom = crud.create_classroom(db, schemas.ClassroomCreate(name=classroom_name))
                    classroom_id = classroom.id
                
                # Create lesson
                lesson = models.Lesson(
                    teacher_id=teacher.id,
                    class_id=school_class.id,
                    classroom_id=classroom_id,
                    subject=subject,
                    day_of_week=day_of_week,
                    lesson_number=lesson_number
                )
                
                db.add(lesson)
                count += 1
                
            except (ValueError, AttributeError) as e:
                continue
        
        db.commit()
        return count
        
    except ET.ParseError as e:
        raise Exception(f"Error parsing XML: {str(e)}")
    except Exception as e:
        raise Exception(f"Error processing XML: {str(e)}")


def parse_plan_html_folder_and_save(html_files: dict, db: Session) -> int:
    """
    Parse HTML files from a folder/ZIP archive containing school schedule.
    
    Args:
        html_files: Dictionary of {filename: file_contents}
        db: Database session
        
    Returns:
        Number of lessons processed
    """
    try:
        count = 0
        
        for filename, contents in html_files.items():
            try:
                soup = BeautifulSoup(contents, 'html.parser')
                
                # Extract class name from filename (e.g., o1.html -> O1)
                class_name = filename.replace('.html', '').replace('.htm', '').upper()
                
                # Get or create school class
                school_class = crud.get_class_by_name(db, class_name)
                if not school_class:
                    school_class = crud.create_class(db, schemas.SchoolClassCreate(name=class_name))
                
                # Find all table rows
                rows = soup.find_all('tr')
                
                for row in rows[1:]:  # Skip header
                    cells = row.find_all(['td', 'th'])
                    if len(cells) < 3:
                        continue
                    
                    try:
                        # Parse columns - adjust based on actual HTML structure
                        # Expected: LessonNumber, Day(or Times), TeacherName, Subject, Classroom
                        
                        lesson_num_text = cells[0].get_text(strip=True)
                        lesson_num = int(''.join(filter(str.isdigit, lesson_num_text)) or 1)
                        
                        # Try to find day information
                        day_of_week = 1  # Default to Monday
                        day_text = cells[1].get_text(strip=True).lower() if len(cells) > 1 else ''
                        
                        # Parse day if it contains Polish day names
                        day_map = {
                            'poniedziałek': 1, 'pon': 1,
                            'wtorek': 2, 'wt': 2,
                            'środa': 3, 'śr': 3,
                            'czwartek': 4, 'czw': 4,
                            'piątek': 5, 'pt': 5
                        }
                        for day_name, day_num in day_map.items():
                            if day_name in day_text:
                                day_of_week = day_num
                                break
                        
                        # Get teacher and subject
                        teacher_name = cells[2].get_text(strip=True) if len(cells) > 2 else ''
                        subject = cells[3].get_text(strip=True) if len(cells) > 3 else ''
                        classroom_name = cells[4].get_text(strip=True) if len(cells) > 4 else None
                        
                        if not teacher_name or not subject:
                            continue
                        
                        # Get or create teacher
                        teacher = crud.get_teacher_by_name(db, teacher_name)
                        if not teacher:
                            short_name = teacher_name[:10]
                            teacher = crud.create_teacher(
                                db,
                                schemas.TeacherCreate(name=teacher_name, short_name=short_name)
                            )
                        
                        # Get or create classroom
                        classroom_id = None
                        if classroom_name and classroom_name != '-':
                            classroom = crud.get_classroom_by_name(db, classroom_name)
                            if not classroom:
                                classroom = crud.create_classroom(db, schemas.ClassroomCreate(name=classroom_name))
                            classroom_id = classroom.id
                        
                        # Create lesson
                        lesson = models.Lesson(
                            teacher_id=teacher.id,
                            class_id=school_class.id,
                            classroom_id=classroom_id,
                            subject=subject,
                            day_of_week=day_of_week,
                            lesson_number=lesson_num
                        )
                        
                        db.add(lesson)
                        count += 1
                        
                    except (ValueError, IndexError):
                        continue
                
                db.commit()
                
            except Exception as e:
                continue
        
        return count
        
    except Exception as e:
        raise Exception(f"Error processing HTML folder: {str(e)}")
